vmm/src/memory_manager.rs

```diff
@@ -1303,8 +1303,10 @@ impl MemoryManager {
             Some(FileOffset::new(f, file_offset))
         } else if let Some(backing_file) = backing_file {
             Some(Self::open_backing_file(backing_file, file_offset, size)?)
-        } else {
+        } else if shared || hugepages {
```

Comment:
I think for hugepages, we shouldn't assume a backing file is required. `mmap()` provides `MAP_HUGETLB` flag in order to allocate using huge pages.
I would apply `MAP_HUGETLB` and `MAP_HUGE_2MB` or `MAP_HUGE_1GB` in case `shared=off` and `hugepages=on`.

Reply:
I thought i'd made it clear in the commit message on the change. You can't combine `MAP_ANONYMOUS` with those options; i.e. the only way to have huge pages is with a file backing it and hence huge pages always requires `MAP_SHARED`. Which as @NunoDasNeves points out means `hugepages=on,shared=off` is contradictory. However i'm loathe to reject it like that since it might break existing scenarios. I would rather redefine `shared` as independent of `hugepages` so as to avoid this issue.